server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/UserEntityValidator.java

```diff
@@ -0,0 +1,84 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.webadmin.service;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import org.apache.james.core.Username;
+
+import com.google.common.collect.ImmutableSet;
+
+public interface UserEntityValidator {
+    class ValidationFailure {
+        private final String errorMessage;
+
+        public ValidationFailure(String errorMessage) {
+            this.errorMessage = errorMessage;
+        }
+
+        public String errorMessage() {
+            return errorMessage;
+        }
+    }
+
+    class EntityType {
+        public static final EntityType GROUP = new EntityType("group");
+        public static final EntityType ALIAS = new EntityType("alias");
+        public static final EntityType USER = new EntityType("user");
+
+        private final String type;
+
+        public EntityType(String type) {
+            this.type = type;
+        }
+
+        public String getType() {
+            return type;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(type);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj instanceof EntityType) {
+                EntityType other = (EntityType) obj;
+                return Objects.equals(this.type, other.type);
+            }
+            return false;
+        }
+    }
+
+    UserEntityValidator NOOP = (username, ignoredTypes) -> Optional.empty();
+
+    static UserEntityValidator aggregate(UserEntityValidator... validators) {
```

Comment:
I don't get why you need this... Can't you use directly `AggregateUserEntityValidator` in your tests instead? Feels wrong to have an implementation of `UserEntityValidator` being returned in a static method of the interface no?

Reply:
> Feels wrong to have an implementation of UserEntityValidator being returned in a static method of the interface no?

Why?

The vararg is convenient. Greater API discovery.

Reply:
Not convinced... API discovery for... only testing? I mean:

`AggregateUserEntityValidator` implements the interface `UserEntityValidator`... But in that interface you have a static method that returns... an `AggregateUserEntityValidator`. I don't think it's a good pattern to do that with an interface? If i'm wrong I really want to understand why, because I don't so far sorry ^^'

Reply:
`Not convinced... API discovery for... only testing?`

Yes.

I don't get why it hurts you.

Reply:
>  But in that interface you have a static method that returns... an AggregateUserEntityValidator. 

Aggregating is IMO a very basic capability inherently related to the working of the interface, and encapsulate no implementation details.